### PR TITLE
Pipeline scope: user can define additional variables 

### LINKED ISSUE #548: Allow the user to define additional symbols in Pipeline scope
The pipeline scope is mainly used for generating nice FPGA code.
In this context, it could be the case that the user needs additional scope-level variables, that are then managed by him itself.

A possible solution could be to have an additional parameter in the Pipeline constructor (a dictionary)

### PR DESCRIPTION
**Pull request by TizianoDeMatteis**
- Resolves #548 
- Resolves a bug in cppunparse: if a symbol is already defined, there is no need to re-define it (or perform type inference when needed). Before, this was causing the re-definition of already defined variables.

### CODE REVIEW

File: dace/codegen/cppunparse.py [hunk 1]
```diff
@@ -320,32 +320,35 @@ def _Assign(self, t):
...
+                        if self.language == dace.dtypes.Language.OpenCL and (
+                                inferred_type is not None
+                                and inferred_type.veclen > 1):
+                            # if the veclen is greater than one, this should be defined with a vector data type
+                            self.write("{}{} ".format(
+                                dace.dtypes._OCL_VECTOR_TYPES[
+                                    inferred_type.type], inferred_type.veclen))
+                        else:
+                            self.write(dace.dtypes._CTYPES[inferred_type.type] +
+                                       " ")
```
Comment (definelicht) [inline dace/codegen/cppunparse.py:348]: How is this related? Maybe add a comment about it on the PR.

File: dace/sdfg/nodes.py [hunk 2]
```diff
@@ -1050,19 +1053,24 @@ class Pipeline(Map):
         dtype=bool,
         default=True,
         desc="Whether to increment regular map indices during pipeline drain.")
+    additional_variables = Property(
+        dtype=dict,
+        desc="Additional variables, managed by the user inside the scope.")
```
Comment (definelicht) [inline dace/sdfg/nodes.py:1058]: Maybe call these `iterators` or `symbols`?

File: dace/sdfg/state.py [hunk 3]
```diff
@@ -1446,6 +1445,10 @@ def add_pipeline(self,
             :param drain_size:    Number of iterations of draining phase.
             :param drain_overlap: Whether the draining phase overlaps with
                                   the "main" streaming phase of the loop.
+            :param additional_variables: a dictionary contanining the additional
+                                  variables that will be created for this scope and that are not
+                                  handled by the scope guards it self.
+                                  The dictionary is in the form 'variable_name' -> init_value
```
Comment (definelicht) [inline dace/sdfg/state.py:1451]:
```suggestion
            :param additional_variables: a dictionary containing additional
                                  symbols that will be created for this scope and that are not
                                  automatically managed by the scope code.
                                  The dictionary takes the form 'variable_name' -> init_value
```

File: tests/fpga/pipeline_scope.py [hunk 4]
```diff
@@ -43,8 +43,8 @@ def make_sdfg(dtype, name="pipeline_test"):
     post_write = post_state.add_write("b")
     post_state.add_memlet_path(post_read,
                                post_write,
-                               memlet=dace.Memlet.simple(post_write,
```
Comment (definelicht) [inline tests/fpga/pipeline_scope.py:46]: Could we take this opportunity to get rid of `Memlet.simple` in this sample? :-)

**Review by definelicht — CHANGES_REQUESTED** [review]

File: dace/codegen/cppunparse.py (hunk 1 above)
Comment (TizianoDeMatteis) [reply dace/codegen/cppunparse.py:348]: Yes. It popped while I was changing the test (but then i changed it again). Will add a comment

**Review by definelicht — APPROVED** [review]
